libcaf_core/src/config_value.cpp

```diff
@@ -126,6 +153,312 @@ const char* config_value::type_name_at_index(size_t index) noexcept {
+        auto ls = element.to_list();
+        if (ls && ls->size() == 2)
+          return tmp.emplace(to_string((*ls)[0]), std::move((*ls)[1])).second;
+        else
+          return false;
+      };
+      if (std::all_of(x.begin(), x.end(), lift)) {
+        return result_type{std::move(tmp)};
+      } else {
+        auto err = make_error(sec::conversion_failed,
+                              "cannot convert list to dictionary unless each "
+                              "element in the list is a key-value pair");
+        return result_type{std::move(err)};
+      }
+    },
+    [](const std::string& x) {
+      if (dictionary tmp; detail::parse(x, tmp) == none) {
+        return result_type{std::move(tmp)};
+      }
+      if (list tmp; detail::parse(x, tmp) == none) {
+        config_value ls{std::move(tmp)};
+        if (auto res = ls.to_dictionary())
+          return res;
+      }
+      std::string msg = "cannot convert ";
+      detail::print_escaped(msg, x);
+      msg += " to a dictionary";
+      return result_type{make_error(sec::conversion_failed, std::move(msg))};
+    },
+    [](const dictionary& x) { return result_type{x}; });
+  return visit(f, data_);
+}
+
+bool config_value::can_convert_to_dictionary() const {
+  auto f = detail::make_overload( //
+    [](const auto&) { return false; },
+    [this](const std::string&) {
+      // TODO: implement some dry-run mode and use it here to avoid creating an
+      //       actual dictionary only to throw it away.
+      auto maybe_dict = to_dictionary();
+      return static_cast<bool>(maybe_dict);
+    },
+    [](const dictionary&) { return true; });
+  return visit(f, data_);
+}
+
+optional<message>
+config_value::parse_msg_impl(string_view str,
+                             span<const type_id_list> allowed_types) {
+  if (auto val = parse(str)) {
+    auto ls_size = val->as_list().size();
+    message result;
+    auto converts = [&val, &result, ls_size](type_id_list ls) {
+      if (ls.size() != ls_size)
+        return false;
+      config_value_reader reader{std::addressof(*val)};
+      auto unused = size_t{0};
+      reader.begin_sequence(unused);
+      CAF_ASSERT(unused == ls_size);
+      intrusive_ptr<detail::message_data> ptr;
+      if (auto vptr = malloc(sizeof(detail::message_data) + ls.data_size()))
+        ptr.reset(new (vptr) detail::message_data(ls), false);
+      else
+        return false;
+      auto pos = ptr->storage();
+      for (auto type : ls) {
+        auto meta = detail::global_meta_object(type);
+        CAF_ASSERT(meta != nullptr);
+        meta->default_construct(pos);
+        ptr->inc_constructed_elements();
+        if (!meta->load(reader, pos))
+          return false;
+        pos += meta->padded_size;
+      }
+      result.reset(ptr.release(), false);
+      return reader.end_sequence();
+    };
+    if (std::any_of(allowed_types.begin(), allowed_types.end(), converts))
+      return {std::move(result)};
```

Comment:
I think warnings should stay reserved for functions with non-obvious pitfalls (like requiring a certain mutex locked for example). FWIW, I think it's obvious that this message needs to perform several memory allocations, otherwise it couldn't return a `message` and the documentation already mentions that it attempts to parse a config-value list.